game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/change/CheckGeneralBattleEnd.java

```diff
@@ -0,0 +1,130 @@
+package games.strategy.triplea.delegate.battle.steps.change;
+
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.DiceRoll;
+import games.strategy.triplea.delegate.ExecutionStack;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import games.strategy.triplea.delegate.battle.steps.BattleStep;
+import games.strategy.triplea.delegate.battle.steps.RetreatChecks;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.AllArgsConstructor;
+
+@AllArgsConstructor
+public class CheckGeneralBattleEnd implements BattleStep {
+  private static final long serialVersionUID = 5172121497955756220L;
+
+  private final BattleState battleState;
+
+  private final BattleActions battleActions;
+
+  protected BattleActions getBattleActions() {
+    return battleActions;
+  }
+
+  protected BattleState getBattleState() {
+    return battleState;
+  }
+
+  @Override
+  public List<String> getNames() {
+    return List.of();
+  }
+
+  @Override
+  public Order getOrder() {
+    return Order.GENERAL_BATTLE_END_CHECK;
+  }
+
+  @Override
+  public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+    if (hasSideLost(BattleState.Side.OFFENSE)) {
+      battleActions.endBattle(bridge);
+      battleActions.defenderWins(bridge);
+
+    } else if (hasSideLost(BattleState.Side.DEFENSE)) {
+      battleActions.endBattle(bridge);
+      battleActions.attackerWins(bridge);
+
+    } else if (isStalemate()) {
+      if (!canAttackerRetreatInStalemate()) {
```

Comment:
Codacy found an issue: [These nested if statements could be combined](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6199427)